src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlClientEventSource.cs

```diff
@@ -300,25 +305,37 @@ internal void TraceEvent<T0>(string message, T0 args0)
         {
             if (Log.IsTraceEnabled())
             {
-                Trace(string.Format(message, args0, args1, args2));
+                Trace(string.Format(message, args0?.ToString(), args1?.ToString(), args2?.ToString()));
             }
         }
 
+        [NonEvent]
+        internal void TraceEventExcluded<T0, T1, T2>(string message, T0 args0, T1 args1, T2 args2)
+        {
+            Trace(string.Format(message, args0?.ToString(), args1?.ToString(), args2?.ToString()));
```

Comment:
In many places, we already pass `string` when we log information. Will additional ToString() calls drop the performance?

Reply:
Those ToStrings are usually not necessary in here as all objects are from parent class of Object  and when we pass them to string.format it should automatically call ToString, but there were some complications when I passed some null objects. it just throws an exception. I added them here to be able to pass null values safely plus when We have requested logs to be captured, I think all the string allocation happen, so there is no difference after that.